pmlc/dialect/tile/transforms/algebraic_opt.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020, Intel Corporation
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/dialect/tile/transforms/passes.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::dialect::tile {
+
+namespace {
+
+// From:
+// %zero = tile.constant(0.0 : f64)
+// %0 = tile.contract add, mul, %zero, %I, %K
+// %1 = tile.add %0, %B
+// %2 = tile.relu %1
+// Into:
+// %0 = tile.contract add, mul, %B, %I, %K
+// %2 = tile.relu %0
+struct AddInitPattern final : public OpRewritePattern<AddOp> {
+  using OpRewritePattern<AddOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AddOp op, PatternRewriter &rewriter) const {
+    return failure(
+        failed(matchAndRewritePermutation(op, rewriter, op.lhs(), op.rhs())) ||
+        failed(matchAndRewritePermutation(op, rewriter, op.rhs(), op.lhs())));
+  }
+
+  LogicalResult matchAndRewritePermutation(AddOp op, PatternRewriter &rewriter,
+                                           Value thisOperand,
+                                           Value otherOperand) const {
+    if (ContractionOp contractOp =
+            dyn_cast_or_null<ContractionOp>(thisOperand.getDefiningOp())) {
```

Comment:
Nit:  Seems like we should cast to contractOp and early return failure if it's null rather than using if just to avoid extra indention and follow LLVM style.